File: TIA/tia_final_project/src/ia/Genetic.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ia;

/**
 *
 * @author crdzbird
 */

import configs.AppConfig;
import java.util.logging.Level;
import java.util.logging.Logger;
import repositories.IndividualImplementation;
import repositories.IndividualInterface;
import repositories.population.PopulationInterface;
import ui.Graph;

public class Genetic<I extends IndividualInterface, P extends PopulationInterface<I>> {

    private static final Logger LOGGER = Logger.getLogger(Genetic.class.getName());
    private static final int REPORT_FREQUENCY = 10;

    private final PopulationInterface<I> population;
    private PopulationInterface<I> memeticPopulation;
    
    private final Graph graph;

    public Genetic(PopulationInterface<I> population, Graph graph) {
        this.population = population;
        this.graph = graph;
    }
    
    

    public Genetic(PopulationInterface<I> population, PopulationInterface<I> populationMemetic, Graph graph) {
        this.population = population;
        this.memeticPopulation = populationMemetic;
        this.graph = graph;
    }

    /**
     * Executes the genetic algorithm for a given number of iterations.
     * 
     * @param iters the number of iterations to run the algorithm for.
     * @return the evolved population after all iterations.
     */
    public PopulationInterface<I> execute(int iters) {
        PopulationInterface<I> p = this.population.copy();

        for (int i = 0; i < iters; i++) {
            evolvePopulation(p, i);
            reportProgress(p, i, iters);
        }

        LOGGER.log(Level.INFO, "\nLAST POPULATION:\n\t{0}", p.retrieveBestIndividual());
        return p;
    }
    
    public PopulationInterface<I> executeWithMemetic(int iters) {
        PopulationInterface<I> p = this.population.copy();
        PopulationInterface<I> mp = this.memeticPopulation.copy();
        
        for (int i = 0; i < iters; i++) {
            evolveBothPopulation(p, mp, i);
            reportProgressWithMemetic(p, mp, i, iters);
        }

        LOGGER.log(Level.INFO, "\nLAST POPULATION:\n\t{0}", p.retrieveBestIndividual());
        return p;
    }

    /**
     * Evolves the population based on given logic.
     *
     * @param p      the population to evolve.
     * @param iteration the current iteration.
     */
    private void evolvePopulation(PopulationInterface<I> p, int iteration) {
        p.evolve(5 % Math.max(1, iteration) == 0);
    }
    
    
    /**
     * Evolves the population based on given logic.
     *
     * @param p      the population to evolve.
     * @param iteration the current iteration.
     */
    private void evolveBothPopulation(PopulationInterface<I> p, PopulationInterface<I> mp, int iteration) {
        p.evolve(5 % Math.max(1, iteration) == 0);
        mp.evolve(5 % Math.max(1, iteration) == 0);
        if (iteration % AppConfig.MEME_POP_RATE == 0) {
                applyMemeticOperations(p, AppConfig.TEMPERATURE, AppConfig.COOLING_RATE);
            }
    }

    /**
     * Reports the progress of the algorithm and updates the graph.
     *
     * @param p        the current population.
     * @param iteration the current iteration.
     * @param total    the total number of iterations.
     */
    private void reportProgress(PopulationInterface<I> p, int iteration, int total) {
        if (iteration % REPORT_FREQUENCY == 0 || iteration == total - 1) {
            IndividualInterface best = p.retrieveBestIndividual();
            LOGGER.log(Level.INFO, "\nIteration: {0}/{1}\t{2}", new Object[]{iteration, total, best});
            graph.addData(iteration, best.getFitness(), 0);
        }
    }
    
    
    private void applyMemeticOperations(PopulationInterface<I> p, double temperature, double coolingRate) {
        IndividualInterface best = p.retrieveBestIndividual();
        IndividualInterface other = p.getMembers().get(AppConfig.RANDOM.nextInt(p.getMembers().size()));

        IndividualInterface bestAfterSA = (I) (new Simulated<>(best, temperature, coolingRate)).start();
        IndividualInterface otherAfterSA = (I) (new Simulated<>(other, temperature, coolingRate)).start();

        p.getMembers().remove(best);
        p.getMembers().remove(other);

        p.getMembers().add(0, (I) bestAfterSA);
        p.getMembers().add((I) otherAfterSA);
    }
    
    
    /**
     * Reports the progress of the algorithm and updates the graph.
     *
     * @param p        the current population.
     * @param iteration the current iteration.
     * @param total    the total number of iterations.
     */
    private void reportProgressWithMemetic(PopulationInterface<I> p, PopulationInterface<I> mp, int iteration, int total) {
        if (iteration % REPORT_FREQUENCY == 0 || iteration == total - 1) {
            IndividualInterface best = p.retrieveBestIndividual();
            IndividualInterface bestMP = mp.retrieveBestIndividual();
            LOGGER.log(Level.INFO, "\nIteration: {0}/{1}\tGenetic:{2}\tMemetic:{3}", new Object[]{iteration, total, best, bestMP});
            graph.addData(iteration, best.getFitness(), bestMP.getFitness());
        }
    }
}

File: TIA/tia_genetic_memetic/lib/algoritms/genetic.dart
import 'dart:math';

import 'package:tia_genetic_memetic/repositories/individual/individual_interface.dart';
import 'package:tia_genetic_memetic/repositories/population/population_interface.dart';

class Genetic<I extends IndividualInterface, P extends PopulationInterface<I>> {
  final P population;
  P? memeticPopulation; // This is nullable, as it seems optional in the Java code

  Genetic(this.population, {this.memeticPopulation}); // Constructor

  P execute(int iters) {
    P p = population.copy()
        as P; // Assuming a copy method exists and works as in Java

    for (int i = 0; i < iters; i++) {
      p.evolve(5 % (1 + i).abs() ==
          0); // Placeholder for evolution logic, replace with actual method call
      reportProgress(p, i, iters); // Assuming you want to keep similar logging
    }

    print(
        'LAST POPULATION: ${p.retrieveBestIndividual()}'); // Simplified logging, consider a better option
    return p;
  }

  P executeWithMemetic(int iters) {
    P p = population.copy() as P;
    P mp = memeticPopulation!.copy()
        as P; // Using the non-null assertion operator, as memeticPopulation is nullable

    for (int i = 0; i < iters; i++) {
      p.evolve(5 % (1 + i).abs() ==
          0); // Placeholder for evolution logic, replace with actual method call
      mp.evolve(5 % (1 + i).abs() == 0); // Same for the memetic population
      reportProgressWithMemetic(
          p, mp, i, iters); // Assuming you want to keep similar logging
    }

    print(
        'LAST POPULATION: ${p.retrieveBestIndividual()}'); // Simplified logging, consider a better option
    return p;
  }

  void reportProgress(
      PopulationInterface<I> p, int iteration, int totalIterations) {
    // This method can be used to report progress of the algorithm.
    // Implementation can vary depending on how you want to report this progress.
    // For simplicity, we'll print the current iteration and the best individual's fitness.
    print(
        'Iteration ${iteration + 1} of $totalIterations: Best Fitness = ${p.retrieveBestIndividual().getFitness()}');
  }

  void reportProgressWithMemetic<T>(PopulationInterface<I> p,
      PopulationInterface<I> mp, int iteration, int total) {
    // Assuming Graph is a class you've defined that has an addData method
    if (iteration % 10 == 0 || iteration == total - 1) {
      IndividualInterface best = p.retrieveBestIndividual();
      IndividualInterface bestMP = mp.retrieveBestIndividual();
      print('Iteration: $iteration/$total\tGenetic: $best\tMemetic: $bestMP');
    }
  }
}
